feat: Preserve user-managed fields when creating namespace

dlipovetsky opened the pull request:
**What problem does this PR solve?**:
We want to ensure the namespace exists, while preserving user-managed fields (i.e. labels and annotations) namespace exists.

Currently, if the namespace already exists, we do not update it. This prevents us from applying our own labels and annotations to an existing namespace, something we need to do in certain cases. This PR allows us to, in a future PR, apply labels and annotations to an existing namespace.

**Which issue(s) this PR fixes**:
Fixes #

**How Has This Been Tested?**:
<!--
Please describe the tests that you ran to verify your changes.
Provide output from the tests and any manual steps needed to replicate the tests.
-->
I added a new integration (envtest-based) test that verifies the desired behavior.

**Special notes for your reviewer**:
<!--
Use this to provide any additional information to the reviewers.
This may include:
- Best way to review the PR.
- Where the author wants the most review attention on.
- etc.
-->
Depends on #556.

dkoshkin commented on pkg/handlers/generic/lifecycle/utils/utils_suite_test.go:
```diff
@@ -0,0 +1,16 @@
+	. "github.com/onsi/gomega"
+)
+
+func TestAMIPatch(t *testing.T) {
```
Did you forget to change the name?

dlipovetsky replied:
Are you suggesting I copy-pasted this and forgot to change the name? How dare you! :joy:

dlipovetsky replied:
Thanks for catching this. I fixed the function, and moved the integration tests to their own file to decrease confusion with the unit tests.

jimmidyson commented on pkg/handlers/generic/lifecycle/utils/utils_integration_test.go:
```diff
@@ -0,0 +1,62 @@
+		namespaceName := "new"
+
+		Expect(EnsureNamespace(ctx, c, namespaceName)).To(Succeed())
+		Expect(c.Delete(ctx, &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: namespaceName,
+			},
+		})).To((Succeed()))
```
Deleting namespaces in envtest does not work due to missing kube controllers. Let's remove this and ensure we use a new namespace name for each spec.

```suggestion
```

dlipovetsky replied:
Thanks. I think you fixed this in https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/pull/559/commits/07543a8204a96403a35bf24bc10cd534642c5d26, which I'll append here,

dlipovetsky commented:
I've rebased on #556, added https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/pull/559/commits/07543a8204a96403a35bf24bc10cd534642c5d26, and force-pushed

dlipovetsky commented:
#556 merged, and had an additional minor fixup, so I've rebased on main and force-pushed.

dkoshkin commented on pkg/handlers/generic/lifecycle/utils/utils.go:
```diff
@@ -105,12 +105,18 @@ func EnsureNamespace(ctx context.Context, c ctrlclient.Client, name string) erro
-	err := client.ServerSideApply(ctx, c, ns, client.ForceOwnership)
+// EnsureNamespace will create the namespace if it does not exist.
+func EnsureNamespace(ctx context.Context, c ctrlclient.Client, ns *corev1.Namespace) error {
+	if ns.TypeMeta.APIVersion == "" {
+		ns.TypeMeta.APIVersion = corev1.SchemeGroupVersion.String()
+	}
+	if ns.TypeMeta.Kind == "" {
+		ns.TypeMeta.Kind = "Namespace"
+	}
```
nit: we can use Go 1.22 magic :) 
```suggestion
	ns.TypeMeta.APIVersion = cmp.Or(ns.TypeMeta.APIVersion, corev1.SchemeGroupVersion.String())
	ns.TypeMeta.Kind = cmp.Or(ns.TypeMeta.APIVersion, "Namespace")
```

jimmidyson replied:
Oh I like this!

dlipovetsky replied:
I appreciate the suggestion. It is more concise, but I find it harder to understand, because you have to (a) know `cmp.Or` semantics, and (b) the zero values of the fields.

Apparently, `cmp.Or` [originated](https://blog.carlana.net/post/2024/golang-cmp-or-uses-and-history/) with a proposal for a `strings.First` function, which seems easier to understand.
